numpyro/contrib/stochastic_support/dcc.py

```diff
@@ -3,6 +3,9 @@
 
 from abc import ABC, abstractmethod
 from collections import OrderedDict, namedtuple
+from typing import Any, Callable, OrderedDict as OrderedDictType
+
+from jaxlib.xla_extension import ArrayImpl
```

Comment:
Regarding the default typing, what are the implications of using `ArrayImpl` vs `jax.numpy.ndarray`, for example?

Reply:
This is a great question! Thanks for the feedback! So we started trying with type-hints with `ArrayImpl` in https://github.com/pyro-ppl/numpyro/pull/1819#discussion_r1649747153 

On the other hand in https://jax.readthedocs.io/en/latest/jax.typing.html its suggested to use `jax.typing.ArrayLike`. However when I use this for `x: ArrayLike` and have inside the function `x.ndim`, `mypy` complains with 

```python
error: Item "builtins.bool" of "Array | ndarray[Any, Any] | numpy.bool | number[Any] | builtins.bool | int | float | complex" has no attribute "ndim"  [union-attr]
```

Hence, this is why I am continuing with `ArrayImpl` and waiting for feedback :)

Reply:
I would like to avoid importing stuff from jaxlib. I guess it is better to use `jax.Array`

Reply:
Re ndim: you can use jnp.ndim(x), which will accept ArrayLike instances, including python scalars.

Reply:
Great input! I'll try that then 🙂!